sycl/doc/extensions/experimental/sycl_ext_codeplay_kernel_fusion.asciidoc

```diff
@@ -0,0 +1,678 @@
+```
+
+
+Table 1. Constructors and destructors of the `fusion_wrapper` class
+|===
+|Constructor|Description
+
+|`explicit fusion_wrapper(queue& syclQueue)`
+
+|Wraps the queue `syclQueue` with a `fusion_wrapper` to get access to the
+fusion API and manage kernel fusion on `syclQueue`. 
+
+The underlying `queue` must have property 
+`sycl::ext::codeplay::property::queue::enable_fusion`
+
+|===
+
+Table 2. Member functions of the `fusion_wrapper` class
+
+|===
+|Member Function|Description
+
+|`void start_fusion()`
+
+|Set the wrapped `queue` into "fusion mode". Subsequent command group
+submissions to the `queue` will not be submitted for execution right away, but rather added to
+a list of kernels that should be fused (i.e., to the _fusion list_), until
+`complete_fusion` or `cancel_fusion` are called.
+
+If the wrapped `queue` is already in fusion mode, the function throws an
+`exception` with `errc::invalid` error code.
+
+|`event complete_fusion(const property_list &)` 
+
+|Complete the fusion: If the
+runtime decides to perform fusion, it will JIT-compile a fused kernel from all
+kernels submitted to the wrapped `queue` since the last call to `start_fusion`
+and submit the fused kernel for execution. Inside the fused kernel, the
+per-work-item effects are executed in the same order as the kernels were
+initially submitted, adding group barriers between each of them by default.
+Otherwise, the individuals kernels will be passed to the scheduler and executed
+in the same order as they were initially submitted. Constraints on when fusion
+is possible and criteria for the implementation to perform fusion are
+implementation-defined. Calling `fusion_wrapper::complete_fusion` does therefore
+not guarantee that the kernels will be fused.
+
+The call is asynchronous, i.e., it may return after fusion (JIT-compilation) is
+done, but before execution of the fused kernel is completed. The returned event
+allows to synchronize with the execution of the fused kernel.
+
+At call completion the wrapped `queue` is no longer in fusion mode, until the
+next `start_fusion`.
+
+|`void cancel_fusion()` 
+
+|Cancel the fusion and submit all kernels submitted to the wrapped `queue` since
+the last `start_fusion()` for immediate execution **without** fusion. The
+kernels are submitted in the same order as they were initially submitted to the
+queue.
+
+This operation is asynchronous, i.e., it may return after the kernels have been
+added to the scheduler, but before any of the previously submitted kernel starts
+or completes execution.
+
+At call completion the wrapped `queue` is no longer in fusion mode, until the next
+`start_fusion`.
+
+|`bool is_in_fusion_mode() const`
+|Returns true if the wrapped SYCL `queue` is currently in fusion mode.
+
+|===
+
+==== Properties
+
+Next to the new API functions and classes described above, this extension also
+adds new properties that are described in Table 3.
+
+Table 3. New properties for kernel fusion.
```

Comment:
These properties should all be in the experimental namespace `sycl::ext::codeplay::experimental::property`.